File: iOS/DittoPOS/Data/Order.swift
//
//  Order.swift
//  DittoPOS
//
//  Copyright © 2026 DittoLive Incorporated. All rights reserved.
//

import Foundation

struct Order: Hashable, Codable {
    let documentId: DocumentID
    var cart: [String: CartLineItem]
    var payments: [String: Payment]
    var statusLog: [String: String]      // iso-timestamp → OrderStatus.rawValue
    var createdAt: Date

    private enum CodingKeys: String, CodingKey {
        case documentId = "_id"
        case cart, payments, createdAt
        case statusLog = "status_log"
    }

    var title: String { String(documentId.id.prefix(8)) }

    var status: OrderStatus { StatusLogDerivation.currentStatus(from: statusLog) }

    var isPaid: Bool {
        status == .canceled || !payments.isEmpty
    }

    static let collectionName = "pos_orders"
}

// MARK: - Construction
extension Order {
    static func new(
        locationId: String,
        createdAt: Date = Date(),
        status: OrderStatus = .open
    ) -> Order {
        let entry = StatusLogDerivation.entry(status, at: createdAt)
        return Order(
            documentId: DocumentID(id: UUID().uuidString, locationId: locationId),
            cart: [:],
            payments: [:],
            statusLog: [entry.timestamp: entry.status],
            createdAt: createdAt
        )
    }
}

// MARK: - Mutation helpers
extension Order {
    func addingCartLineItem(_ lineItem: CartLineItem, lineItemId: String) -> Order {
        var copy = self
        copy.cart[lineItemId] = lineItem
        let entry = StatusLogDerivation.entry(.inProcess)
        copy.statusLog[entry.timestamp] = entry.status
        return copy
    }

    func addingPayment(_ payment: Payment, paymentId: String) -> Order {
        var copy = self
        copy.payments[paymentId] = payment
        return copy
    }

    func appendingStatus(_ status: OrderStatus, at date: Date = Date()) -> Order {
        var copy = self
        let entry = StatusLogDerivation.entry(status, at: date)
        copy.statusLog[entry.timestamp] = entry.status
        return copy
    }
}

// MARK: - Derived views
extension Order {
    var sortedLineItems: [CartLineItem] {
        cart.values.sorted { $0.createdAt < $1.createdAt }
    }

    var totalCents: Int {
        sortedLineItems.reduce(0) { $0 + $1.price.amount * $1.qty }
    }

    var total: Price { Price(cents: totalCents) }

    var summary: [String: Int] {
        var out: [String: Int] = [:]
        for line in sortedLineItems {
            out[line.name, default: 0] += line.qty
        }
        return out
    }
}

// MARK: - Preview
extension Order {
    static func preview() -> Order {
        Order.new(locationId: "PreviewLocationId")
    }
}

File: Android/app/src/main/java/live/ditto/pos/core/data/orders/Order.kt
package live.ditto.pos.core.data.orders

import kotlinx.datetime.Clock
import kotlinx.datetime.Instant
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import live.ditto.pos.core.data.CartLineItem
import live.ditto.pos.core.data.DittoInstantSerializer
import live.ditto.pos.core.data.DocumentID
import live.ditto.pos.core.data.OrderStatus
import live.ditto.pos.core.data.Payment
import live.ditto.pos.core.data.Price
import live.ditto.pos.core.data.StatusLogDerivation
import java.util.UUID

@Serializable
data class Order(
    @SerialName("_id") val documentId: DocumentID,
    val cart: Map<String, CartLineItem> = emptyMap(),
    val payments: Map<String, Payment> = emptyMap(),
    @SerialName("status_log") val statusLog: Map<String, String> = emptyMap(),
    @Serializable(with = DittoInstantSerializer::class)
    val createdAt: Instant
) {
    val title: String get() = documentId.id.take(8)

    val status: OrderStatus get() = StatusLogDerivation.currentStatus(statusLog)
    val isPaid: Boolean get() = status == OrderStatus.CANCELED || payments.isNotEmpty()

    val sortedLineItems: List<CartLineItem>
        get() = cart.values.sortedBy { it.createdAt }

    val totalCents: Int get() = sortedLineItems.sumOf { it.price.amount * it.qty }
    val total: Price get() = Price(totalCents)

    /** name → quantity, used by KDS summary view. */
    val summary: Map<String, Int>
        get() = sortedLineItems.groupingBy { it.name }.fold(0) { acc, line -> acc + line.qty }

    fun addingCartLineItem(lineItem: CartLineItem, lineItemId: String): Order {
        val (timestamp, statusValue) = StatusLogDerivation.entry(OrderStatus.IN_PROCESS)
        return copy(
            cart = cart + (lineItemId to lineItem),
            statusLog = statusLog + (timestamp to statusValue)
        )
    }

    fun addingPayment(payment: Payment, paymentId: String): Order =
        copy(payments = payments + (paymentId to payment))

    fun appendingStatus(newStatus: OrderStatus, at: Instant = Clock.System.now()): Order {
        val (timestamp, statusValue) = StatusLogDerivation.entry(newStatus, at = at)
        return copy(statusLog = statusLog + (timestamp to statusValue))
    }

    companion object {
        const val COLLECTION_NAME = "pos_orders"

        fun new(
            locationId: String,
            createdAt: Instant = Clock.System.now(),
            status: OrderStatus = OrderStatus.OPEN
        ): Order {
            val (timestamp, statusValue) = StatusLogDerivation.entry(status, at = createdAt)
            return Order(
                documentId = DocumentID(id = UUID.randomUUID().toString(), locationId = locationId),
                cart = emptyMap(),
                payments = emptyMap(),
                statusLog = mapOf(timestamp to statusValue),
                createdAt = createdAt
            )
        }
    }
}
